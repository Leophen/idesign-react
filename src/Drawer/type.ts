export interface DrawerProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 控制抽屉显示隐藏
   * @default false
   */
  visible?: boolean;
  /**
   * 按下退出键是否触发关闭事件
   * @default true
   */
  closeOnEsc?: boolean;
  /**
   * 是否显示遮罩层
   * @default true
   */
  showMask?: boolean;
  /**
   * 抽屉展开位置
   * @default right
   */
  placement?: 'left' | 'right' | 'top' | 'bottom'
  /**
   * 抽屉宽度
   */
  width?: React.CSSProperties["width"];
  /**
   * 抽屉高度
   */
  height?: React.CSSProperties["height"];
  /**
   * 抽屉头部内容
   */
  header?: React.ReactNode;
  /**
   * 抽屉底部内容
   */
  footer?: React.ReactNode;
  /**
   * 抽屉关闭时触发事件
   */
  onClose?: () => void;
}
