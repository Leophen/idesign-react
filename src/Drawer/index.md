---
nav:
  title: 组件
  path: /components
group:
  title: 弹框提示组件
  order: 5
order: 5
---

# Drawer 抽屉

屏幕边缘滑出的抽屉式浮层面板。

## 基本用法

可通过 `visible` 属性控制抽屉显示隐藏。

```tsx
import React, { useState } from 'react';
import { Button, Drawer } from 'idesign-react';

const App = () => {
  const [visible, setVisible] = useState(false);
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible(true)}>
          打开抽屉
        </Button>
        <Drawer header="抽屉标题" visible={visible} onClose={() => setVisible(false)}>
          <span>抽屉内容</span>
        </Drawer>
      </div>
    </div>
  );
};

export default App;
```

## 禁用退出键关闭功能

可通过 `closeOnEsc` 属性控制打开抽屉时按下 ESC 退出键是否触发关闭事件，默认为 `true`，表示 ESC 退出键触发 `onClose` 关闭事件。

```tsx
import React, { useState } from 'react';
import { Button, Drawer } from 'idesign-react';

const App = () => {
  const [visible, setVisible] = useState(false);
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible(true)}>
          禁用抽屉退出键功能
        </Button>
        <Drawer
          header="抽屉标题"
          visible={visible}
          closeOnEsc={false}
          onClose={() => setVisible(false)}
        >
          <span>抽屉内容</span>
        </Drawer>
      </div>
    </div>
  );
};

export default App;
```

## 隐藏遮罩层

可通过 `showMask` 属性控制打开抽屉时是否显示遮罩层，默认为 `true`。

```tsx
import React, { useState } from 'react';
import { Button, Drawer } from 'idesign-react';

const App = () => {
  const [visible, setVisible] = useState(false);
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible(true)}>
          隐藏遮罩层
        </Button>
        <Drawer
          header="抽屉标题"
          visible={visible}
          showMask={false}
          onClose={() => setVisible(false)}
        >
          <span>抽屉内容</span>
        </Drawer>
      </div>
    </div>
  );
};

export default App;
```

## 自定义抽屉方向

可通过 `placement` 属性控制抽屉展开方向，默认为 `right`。

```tsx
import React, { useState } from 'react';
import { Button, Drawer } from 'idesign-react';

const App = () => {
  const [visible1, setVisible1] = useState(false);
  const [visible2, setVisible2] = useState(false);
  const [visible3, setVisible3] = useState(false);
  const [visible4, setVisible4] = useState(false);

  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible1(true)}>
          默认方向
        </Button>
        <Button type="primary" onClick={() => setVisible2(true)}>
          从左展开
        </Button>
        <Button type="primary" onClick={() => setVisible3(true)}>
          从上展开
        </Button>
        <Button type="primary" onClick={() => setVisible4(true)}>
          从下展开
        </Button>
        <Drawer header="抽屉标题" visible={visible1} onClose={() => setVisible1(false)}>
          <span>抽屉内容</span>
        </Drawer>
        <Drawer
          header="抽屉标题"
          visible={visible2}
          placement="left"
          onClose={() => setVisible2(false)}
        >
          <span>抽屉内容</span>
        </Drawer>
        <Drawer
          header="抽屉标题"
          visible={visible3}
          placement="top"
          onClose={() => setVisible3(false)}
        >
          <span>抽屉内容</span>
        </Drawer>
        <Drawer
          header="抽屉标题"
          visible={visible4}
          placement="bottom"
          onClose={() => setVisible4(false)}
        >
          <span>抽屉内容</span>
        </Drawer>
      </div>
    </div>
  );
};

export default App;
```

## 自定义抽屉宽高

可通过 `width` 或 `height` 属性控制抽屉宽高。

```tsx
import React, { useState } from 'react';
import { Button, Drawer } from 'idesign-react';

const App = () => {
  const [visible1, setVisible1] = useState(false);
  const [visible2, setVisible2] = useState(false);
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible1(true)}>
          自定义宽度
        </Button>
        <Button type="primary" onClick={() => setVisible2(true)}>
          自定义高度
        </Button>
        <Drawer header="抽屉标题" visible={visible1} width={500} onClose={() => setVisible1(false)}>
          <span>抽屉内容</span>
        </Drawer>
        <Drawer
          header="抽屉标题"
          visible={visible2}
          placement="top"
          height={500}
          onClose={() => setVisible2(false)}
        >
          <span>抽屉内容</span>
        </Drawer>
      </div>
    </div>
  );
};

export default App;
```

## 自定义抽屉头部

可通过 `header` 属性控制抽屉头部内容。

```tsx
import React, { useState } from 'react';
import { Button, Drawer, Icon } from 'idesign-react';

const header = (
  <>
    <Icon name="Histogram" />
    <span>这是一个自定义标题</span>
  </>
);

const App = () => {
  const [visible, setVisible] = useState(false);
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible(true)}>
          自定义头部
        </Button>
        <Drawer header={header} visible={visible} onClose={() => setVisible(false)}>
          <span>抽屉内容</span>
        </Drawer>
      </div>
    </div>
  );
};

export default App;
```

## 自定义抽屉底部

可通过 `footer` 属性控制抽屉底部内容，当设为 `false` 时隐藏抽屉底部内容。

```tsx
import React, { useState } from 'react';
import { Button, Drawer, Icon } from 'idesign-react';

const footer = (
  <>
    <Button variant="outline" onClick={() => setVisible1(false)}>
      我再想想
    </Button>
    <Button>去意已决</Button>
  </>
);

const App = () => {
  const [visible1, setVisible1] = useState(false);
  const [visible2, setVisible2] = useState(false);

  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button type="primary" onClick={() => setVisible1(true)}>
          自定义底部
        </Button>
        <Button type="primary" onClick={() => setVisible2(true)}>
          隐藏抽屉底部
        </Button>
        <Drawer
          header="抽屉标题"
          footer={footer}
          visible={visible1}
          onClose={() => setVisible1(false)}
        >
          <span>抽屉内容</span>
        </Drawer>
        <Drawer
          header="抽屉标题"
          footer={false}
          visible={visible2}
          onClose={() => setVisible2(false)}
        >
          <span>抽屉内容</span>
        </Drawer>
      </div>
    </div>
  );
};

export default App;
```

<API />
