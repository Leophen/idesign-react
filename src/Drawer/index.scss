.i-drawer-wrapper {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 2400;
  width: 100vw;
  height: 100vh;
  pointer-events: none;
}

.i-drawer__mask {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 2400;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.6);
  pointer-events: auto;
}

.i-drawer {
  position: absolute;
  top: 0;
  right: 0;
  z-index: 2401;
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  width: 360px;
  height: 100%;
  overflow: auto;
  color: var(--i-font-2);
  font-size: 14px;
  background: var(--i-bg);
  border-left: 1px solid var(--i-border);
  box-shadow: var(--i-shadow);
  pointer-events: auto;
  &.i-drawer--placement-left {
    right: unset;
    left: 0;
    border-left: unset;
    border-right: 1px solid var(--i-border);
  }
  &.i-drawer--placement-top {
    width: 100%;
    height: 300px;
    border-left: unset;
    border-bottom: 1px solid var(--i-border);
  }
  &.i-drawer--placement-bottom {
    top: unset;
    bottom: 0;
    width: 100%;
    height: 300px;
    border-left: unset;
    border-top: 1px solid var(--i-border);
  }
}

.i-drawer__close {
  position: absolute;
  top: 14px;
  right: 14px;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 50%;
  cursor: pointer;
  transition: all 0.1s ease-out;
  .i-icon {
    margin: 0;
  }
  &:hover {
    background: var(--i-bg-hover);
  }
}

.i-drawer__header,
.i-drawer__body,
.i-drawer__footer {
  box-sizing: border-box;
  padding: 16px;
}

.i-drawer__header {
  display: flex;
  align-items: center;
  font-weight: 700;
  font-size: 16px;
  word-break: break-word;
  border-bottom: 1px solid var(--i-border);
  .i-icon {
    margin-right: 8px;
    font-size: 20px;
  }
}

.i-drawer__body {
  flex: 1;
  overflow: auto;
  overflow: overlay;
}

.i-drawer__footer {
  width: 100%;
  text-align: right;
  border-top: 1px solid var(--i-border);
  .i-button + .i-button {
    margin-left: 10px;
  }
}
