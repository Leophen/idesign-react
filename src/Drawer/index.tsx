import React, { useEffect, useState, useRef } from 'react';
import classNames from 'classnames';
import './index.scss';
import Icon from '../Icon';
import Button from '../Button';
import Transition from '../Transition';
import ReactDOM from 'react-dom';
import { DrawerProps } from './type';

// 获取触发抽屉打开的 DOM 节点原位置
let clickOpenTarget: EventTarget | null;
const getClickPosition = (e: MouseEvent) => {
  clickOpenTarget = e.target
};
if (typeof window !== 'undefined' && window.document && window.document.documentElement) {
  document.documentElement.addEventListener('click', getClickPosition, true);
}

// 创建抽屉容器
let drawerWrapper = document.querySelector('#i-drawer-wrapper')
if (!drawerWrapper) {
  drawerWrapper = document.createElement('div')
  drawerWrapper.className = 'i-drawer-wrapper'
  drawerWrapper.id = 'i-drawer-wrapper'
  document.body.append(drawerWrapper)
}

const Drawer: React.FC<DrawerProps> = (props) => {
  const {
    children = '',
    className,
    style,
    visible = false,
    closeOnEsc = true,
    showMask = true,
    placement = 'right',
    width,
    height,
    header,
    footer,
    onClose = () => { },
    ...restProps
  } = props;

  const handleKeyDown = (e: any) => {
    if (e.key === 'Escape') {
      onClose?.()
    }
  };

  const closeDrawer = () => {
    onClose?.()
    closeOnEsc && document.removeEventListener('keydown', handleKeyDown)
  }

  // 判断点击节点是否父元素内
  const hasParent = (node: any, parent: HTMLElement) => {
    while (node) {
      if (node === parent) {
        return true;
      }
      node = node.parentNode
    }
    return false;
  };

  const handleClick = (e: any) => {
    if (!hasParent(e.target, drawer.current) && e.target !== clickOpenTarget) {
      closeDrawer()
      document.removeEventListener('click', handleClick, true)
    }
  }

  const drawer = useRef<any>(null);

  // 打开抽屉时禁止背景滚动，对原 overflow 进行备份
  const bodyOverflow = useRef<string>(document.body.style.overflow);

  useEffect(() => {
    if (visible) {
      // 打开抽屉时禁止背景滚动
      document.body.style.overflow = 'hidden';
      // 退出键功能
      closeOnEsc && document.addEventListener('keydown', handleKeyDown)
      // 无遮罩层时点击关闭功能
      !showMask && document.addEventListener('click', handleClick, true)
    } else {
      // 关闭抽屉时恢复背景滚动
      document.body.style.overflow = bodyOverflow.current;
    }
  }, [visible])

  const drawerNode = (
    <>
      {showMask && <Transition
        timeout={200}
        in={visible}
        animation='fade-in'
        key='i-drawer__mask'
      >
        <div className="i-drawer__mask" onClick={closeDrawer}></div>
      </Transition>}

      <Transition
        timeout={200}
        in={visible}
        animation={`slide-in-${placement}`}
        key='i-drawer'
      >
        <div
          ref={drawer}
          className={classNames(
            'i-drawer',
            placement !== 'right' && `i-drawer--placement-${placement}`,
            className
          )}
          style={{ ...(style || {}), ...{ width: width, height: height } }}
          {...restProps}
        >
          <div className="i-drawer__close" onClick={closeDrawer}>
            <Icon name="Close" />
          </div>
          <div className="i-drawer__header">{header}</div>
          <div className="i-drawer__body">{children}</div>
          {footer !== false ?
            <div className="i-drawer__footer">
              {footer ? footer : (
                <>
                  <Button variant="outline" onClick={closeDrawer}>取消</Button>
                  <Button>确认</Button>
                </>
              )}
            </div> : <></>
          }
        </div>
      </Transition>
    </>
  )

  return (
    ReactDOM.createPortal(drawerNode, drawerWrapper as HTMLElement)
  )
};

Drawer.displayName = 'Drawer';

export default Drawer;
