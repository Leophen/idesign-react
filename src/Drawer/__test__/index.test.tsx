import { render } from '@testing-library/react';
import Drawer from '../index';

describe('Drawer 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(
      <Drawer header="抽屉标题" visible={true}>
        <span>抽屉内容</span>
      </Drawer>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

