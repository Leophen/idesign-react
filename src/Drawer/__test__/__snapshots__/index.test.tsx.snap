// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer 组件测试 create 1`] = `<DocumentFragment />`;
